lib/evaluate/intrinsics.cc

```diff
@@ -0,0 +1,1244 @@
+
+// This file defines the supported intrinsic procedures and implements
+// their recognition and validation.  It is largely table-driven.  See
+// documentation/intrinsics.md and section 16 of the Fortran 2018 standard
+// for full details on each of the intrinsics.  Be advised, they have
+// complicated details, and the design of these tables has to accommodate
+// that complexity.
+
+// Dummy arguments to generic intrinsic procedures are each specified by
+// their keyword name (rarely used, but always defined), allowable type
+// categories, a kind pattern, a rank pattern, and information about
+// optionality and defaults.  The kind and rank patterns are represented
+// here with code values that are significant to the matching/validation engine.
+
+// These are small bit-sets of type category enumerators.
+// Note that typeless (BOZ literal) values don't have a distinct type category.
+// These typeless arguments are represented in the tables as if they were
+// INTEGER with a special "typeless" kind code.  Arguments of intrinsic types
+// that can also be be typeless values are encoded with an "elementalOrBOZ"
+// rank pattern.
+using CategorySet = common::EnumSet<TypeCategory, 8>;
+static constexpr CategorySet Int{TypeCategory::Integer};
+static constexpr CategorySet Real{TypeCategory::Real};
+static constexpr CategorySet Complex{TypeCategory::Complex};
+static constexpr CategorySet Char{TypeCategory::Character};
+static constexpr CategorySet Logical{TypeCategory::Logical};
+static constexpr CategorySet IntOrReal{Int | Real};
+static constexpr CategorySet Floating{Real | Complex};
+static constexpr CategorySet Numeric{Int | Real | Complex};
+static constexpr CategorySet Relatable{Int | Real | Char};
+static constexpr CategorySet IntrinsicType{
+    Int | Real | Complex | Char | Logical};
+static constexpr CategorySet AnyType{
+    IntrinsicType | CategorySet{TypeCategory::Derived}};
+
+ENUM_CLASS(KindCode, none, defaultIntegerKind,
+    defaultRealKind,  // is also the default COMPLEX kind
+    doublePrecision, defaultCharKind, defaultLogicalKind,
+    any,  // matches any kind value; each instance is independent
+    typeless,  // BOZ literals are INTEGER with this kind
+    teamType,  // TEAM_TYPE from module ISO_FORTRAN_ENV (for coarrays)
+    kindArg,  // this argument is KIND=
+    effectiveKind,  // for function results: same "kindArg", possibly defaulted
+    dimArg,  // this argument is DIM=
+    same,  // match any kind; all "same" kinds must be equal
+    likeMultiply,  // for DOT_PRODUCT and MATMUL
+)
+
+struct TypePattern {
+  CategorySet categorySet;
+  KindCode kindCode{KindCode::none};
+  std::ostream &Dump(std::ostream &) const;
+};
+
+// Abbreviations for argument and result patterns in the intrinsic prototypes:
+
+// Match specific kinds of intrinsic types
+static constexpr TypePattern DftInt{Int, KindCode::defaultIntegerKind};
+static constexpr TypePattern DftReal{Real, KindCode::defaultRealKind};
```

Comment:
"Dft" is unfortunate as a prefix, but I guess we already have type named "DefaultReal".

Reply:
Yes.  But that `DefaultReal` type isn't really used and maybe I should remove or rename it.

Reply:
After I commit these changes, I will expunge and replace the types like DefaultReal with a better pervasive solution, and then rename these objects with better names once they're available.